multipart_file_upload.rb

```diff
@@ -0,0 +1,303 @@
+  num_parts += 1 if remainder > 0
+
+  part_number = 1
+  read_bytes = 0
+
+  # Read the file a chunk at a time and write the chunk to S3
+  File.open(file_path) do |file|
+    until file.eof?
+      remaining_bytes = file_size_bytes - read_bytes
+      buffer_size = if remaining_bytes < part_size_bytes
+                      remaining_bytes
+                    else
+                      part_size_bytes
+                    end
+      buffer = file.read(buffer_size)
+
+      url = prepare_file_part(part_number, s3_key, upload_id)
+
+      uri = URI(url)
+      Net::HTTP.start(uri.host) do |http|
+        response = http.send_request('PUT', uri, buffer)
+        puts "upload part #{part_number} response.code = #{response.code}"
+
+        # For production use you would add appropriate
+        # error handling here, such as retrying the upload
+        raise "Error reading part #{part_number}, response code: #{response.code}" unless response.code == '200'
+
+        # eTags are required to be wrapped in double quotes
+        # (https://www.rfc-editor.org/rfc/rfc2616#section-14.19),
+        # but we don't want these because they'll mess up the
+        # GraphQL query we'll issue later on, so remove them
+        etag = response.header['etag']
+        etag = etag[1, etag.length - 2]
+
+        part = {
+          part_number: part_number,
+          etag: etag
+        }
+        parts.push(part)
+      end
+
+      part_number += 1
+      read_bytes += buffer_size
+    end
+  end
+
+  parts
+end
+
+#
+# This method demonstrates how to get a pre-signed URL
+# to upload a specific part of a file.
+#
+# @param [string] part_number Part number between 1 and 10,000
+# @param [string] s3_key S3 key of the file
+# @param [string] upload_id ID of the upload
+#
+# @return [string] Pre-signed URL for the part
+#
+def prepare_file_part(part_number, s3_key, upload_id)
+  gql = <<~GQL
+    mutation PrepareMultipartFileUploadPart {
+      prepare_multipart_file_upload_part (
+        part_number: #{part_number}
+        s3_key: "#{s3_key}"
+        upload_id: "#{upload_id}"
+      ) {
+        url
+      }
+    }
+  GQL
+
+  response = make_graphql_request(gql)
+  json = JSON.parse(response.body)
+  results = json.dig('data', 'prepare_multipart_file_upload_part')
+  results['url']
+end
+
+#
+# This methods demonstrates how to complete a multipart
+# file upload. You need to provide the upload ID, the S3 key,
+# a list of parts (part number and e-tag).
+#
+# @param [<Array<Hash>>] parts Array of part information
+# @param [string] s3_key S3 key of the file
+# @param [string] upload_id ID of the upload
+#
+# @return [Boolean] True if request was successful
+#
+def complete_multipart_file_upload(parts, s3_key, upload_id)
+  puts 'Complete multipart file upload'
+
+  parts_gql = '[' + parts.reduce("") do |memo, part|
+    memo += %Q|{part_number: #{part[:part_number]}, etag: "#{part[:etag]}"}|
+    memo
+  end + ']'
```

Comment:
Would it be easier to write a query with a variable so that you didn't have to serialize GQL (could just pass a more normal js array into variable)?

Reply:
Thanks for the suggestion, that cleaned things up nicely! https://github.com/SenteraLLC/file-upload-examples/pull/2/commits/a7cb16b058a554f14ebd94826d6a1812ecf023e7